pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1231,45 +1250,70 @@ public void updateKafkaPartitionsIfNecessary(TableConfig tableConfig) {
+
+    // get current partition assignments for all tables in same tenant
+    Map<String, ZNRecord> currentPartitionAssignment = new HashMap<>(realtimeTablesWithSameTenant.size());
+
+    for (String tableName : realtimeTablesWithSameTenant) {
+      allTableConfigsInTenant.put(tableName, getRealtimeTableConfig(tableName));
+      ZNRecord kafkaPartitionAssignment = getKafkaPartitionAssignment(tableName);
+      if (kafkaPartitionAssignment != null) {
+        currentPartitionAssignment.put(tableName, kafkaPartitionAssignment);
       }
+    }
+    allTableConfigsInTenant.put(tableConfig.getTableName(), tableConfig);
 
-      LinkedHashMap<String, Integer> states = new LinkedHashMap<>(2);
-      states.put("OFFLINE", 0);
-      states.put("ONLINE", nReplicas);
-
-      AutoRebalanceStrategy strategy = new AutoRebalanceStrategy(resourceName, partitions, states);
-      znRecord = strategy.computePartitionAssignment(instanceNames, instanceNames, new HashMap<String, Map<String, String>>(0), null);
-      znRecord.setMapFields(new HashMap<String, Map<String, String>>(0));
-    }
-    */
-    {
-      int serverId = 0;
-      for (int p = 0; p < nPartitions; p++) {
-        List<String> instances = new ArrayList<>(nReplicas);
-        for (int r = 0; r < nReplicas; r++) {
-          instances.add(instanceNames.get(serverId++));
-          if (serverId == instanceNames.size()) {
-            serverId = 0;
-          }
-        }
-        znRecord.setListField(Integer.toString(p), instances);
+    PartitionAssignmentGenerator partitionAssignmentGenerator;
+    if (isPartitionAwareRealtimeRouting(tableConfig)) {
+      partitionAssignmentGenerator =
+          new PartitionAwarePartitionAssignmentGenerator(tableConfig, nKafkaPartitions, instanceNames,
+              allTableConfigsInTenant, currentPartitionAssignment);
+    } else {
+      partitionAssignmentGenerator =
+          new AutoRebalancePartitionAssignmentGenerator(tableConfig, nKafkaPartitions, instanceNames,
+              allTableConfigsInTenant, currentPartitionAssignment);
+    }
+    return partitionAssignmentGenerator.generatePartitionAssignment();
+  }
+
+  /**
+   * Get all realtime tables with given tenant
+   * @param serverTenantName
+   * @return
+   */
+  protected List<String> getRealtimeTablesWithServerTenant(String serverTenantName) {
+    List<String> realtimeTablesWithServerTenant = new ArrayList<>();
+    for (String tableName : _helixResourceManager.getAllRealtimeTables()) {
+      TableConfig realtimeTableConfig = getRealtimeTableConfig(tableName);
+      if (realtimeTableConfig.getTenantConfig().getServer().equals(serverTenantName)) {
+        realtimeTablesWithServerTenant.add(tableName);
       }
     }
-    return znRecord;
+    return realtimeTablesWithServerTenant;
+  }
+
+  /**
+   * Checks if given table has partition aware routing config set
+   * @param tableConfig
+   * @return
+   */
+  public boolean isPartitionAwareRealtimeRouting(TableConfig tableConfig) {
```

Comment:
this logic seems to be duplicated elsewhere. Instead of checking explicitly for the name, its better to have boolean on the RoutingConfig.isPartitionAware right? 
Also why do we care if its partitionAware or not

Reply:
@kishoreg , PartitionAwareRealtimeRoutingTableBuilder attempts to optimize the number of servers to which the request is sent in the case that the number of consuming  hosts is a multiple of number of replicas, AND we do a uniform (kafka/stream) partition allocation strategy (what is referred to as 'spraying' in this PR). The routing table builder will still work if these conditions don't match, except that the request fan-out will be sub-optimal for the high performance use cases where we want to minimize the number of servers participating in a query.

What we are attempting to do here is to continue to preserve that optimized behavior as much as possible. Using helix auto-rebalancer will  throw off partition-aware routing completely. 

We can discuss more offline.